scipy/sparse/sparsetools/coo.h

```diff
@@ -145,4 +145,39 @@ void coo_matvec(const npy_int64 nnz,
     }
 }
 
+
+
+/*
+ *
+ * Input Arguments:
+ *   npy_int64  nnz     - number of nonzeros in A
+ *   npy_int64  n_col_B - number of columns in B
+ *   I  Ai[nnz]         - row indices
+ *   I  Aj[nnz]         - column indices
+ *   T  Ax[nnz]         - nonzero values
+ *   T  Bx[n_col]       - input matrix flattened
+ *
+ * Output Arguments:
+ *   T  Yx[n_row]     - output matrix flattened
+ *
+ * Notes:
+ *   Output array Yx must be preallocated
```

Comment:
This looks like it would replace `coo_matvec` nicely too. Then we would avoid adding a new function -- instead replacing one with another.  But I'm not sure whether the more general `coo_matmul_dense` is slower than `coo_matvec`.  It has an extra `* n_col_B +i` where `n_col_B` would be `1` and `i` would be `0`. I think that would not slow things down too much.  But sometimes my intuition doesn't align with actual results.

Can you check whether replacing `coo_matvec` with `coo_matmat_dense` slows down matvec calculations significantly?

Reply:
I benchmarked matrix-vector multiplication using `coo_matvec` vs using `coo_matmat_dense`, and got the following results -
| shapes                 | iterations | coo_matvec   | coo_matmat_dense |
|:-----------------------|:-----------|:-------------|:-----------------|
| (10000,10000) x (10000,) | 100        | 9.72s, 9.71s | 10.6s, 10.9s     |
| (100,1000) x (1000,)   | 100000     | 3.90s, 3.57s | 5.20s, 5.01s     |
| (100000,100) x (100,)    | 10000      | 8.78s, 8.15s | 10.1s, 9.68s     |

There is a slight difference as expected but I think it doesn't significantly slow down the matrix-vector calculations.

Reply:
Those results seem like a bigger penalty than I would expect. It might be worth looking into where that slowdown is coming from more closely.

Reply:
Although expected (since the dense matrix is in C order so a copy shouldn't be made anyway), I tried benchmarking without `.ravel('C')` but didn't spot any difference in the timings. I don't think any code in `_matmul_dispatch` or any other intermediary functions could be causing the slowdown either. Do you have any speculations?

(Note - for mat-vec multiplication, I replaced `coo_matvec(self.nnz, row, col, self.data, other, result)` with `coo_matmat_dense(self.nnz, 1, row, col, self.data, other.ravel('C'), result)`.)

Reply:
Since it doesn't seem to affect the library size nor build time and the timing is slightly slower with a single function, I think we should support both functions. I'll leave this conversation unresolved in case others have ideas, but I think we should implement it as is and come back to removing `coo_matvec` by using `coo_matmat_dense` in another PR if desired.